pkg/autodiscovery/providers/cloudfoundry.go

```diff
@@ -0,0 +1,206 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2020 Datadog, Inc.
+//
+// +build clusterchecks
+
+package providers
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/autodiscovery/integration"
+	"github.com/DataDog/datadog-agent/pkg/autodiscovery/providers/names"
+	"github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/util/cloudfoundry"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/bhmj/jsonslice"
+)
+
+// CloudFoundryConfigProvider implements the Config Provider interface, it should
+// be called periodically and returns templates from Cloud Foundry BBS for AutoConf.
+type CloudFoundryConfigProvider struct {
+	bbsCache      cloudfoundry.BBSCacheI
+	lastCollected time.Time
+}
+
+// NewCloudFoundryConfigProvider instantiates a new CloudFoundryConfigProvider from given config
+func NewCloudFoundryConfigProvider(conf config.ConfigurationProviders) (ConfigProvider, error) {
+	cfp := CloudFoundryConfigProvider{
+		lastCollected: time.Now(),
+	}
+	var err error
+
+	cfp.bbsCache, err = cloudfoundry.GetGlobalBBSCache()
+	if err != nil {
+		return nil, err
+	}
+	return cfp, nil
+}
+
+// String returns a string representation of the CloudFoundryConfigProvider
+func (cf CloudFoundryConfigProvider) String() string {
+	return names.CloudFoundryBBS
+}
+
+// IsUpToDate returns true if the last collection time was later than last BBS Cache refresh time
+func (cf CloudFoundryConfigProvider) IsUpToDate() (bool, error) {
+	return cf.lastCollected.After(cf.bbsCache.LastUpdated()), nil
+}
+
+// Collect collects AD config templates from all relevant BBS API information
```

Comment:
`from all relevant BBS API information` what do you mean?

Reply:
This means that both desired and actual LRPs come into play (and theoretically we might need more bits and pieces of info in the future).